pkg/controller/appdefinition/deploy.go

```diff
@@ -407,6 +418,65 @@ func toContainer(app *v1.AppInstance, tag name.Reference, deploymentName, contai
 		StartupProbe:   toProbe(container, v1.StartupProbeType),
 		ReadinessProbe: toProbe(container, v1.ReadinessProbeType),
 	}
+
+	resources, err := toResources(req, app, container, containerName)
+	if resources != nil {
+		containerObject.Resources = *resources
+	}
+
+	return containerObject, err
+}
+
+func toResources(req router.Request, app *v1.AppInstance, container v1.Container, containerName string) (*corev1.ResourceRequirements, error) {
+	var memBytes int64
+
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.WorkloadMemoryDefault != nil {
+		memBytes = *cfg.WorkloadMemoryDefault
+	}
+
+	// Determine which memory should be used to set the resource limit/requests.
+	if m, set := app.Spec.Memory.Workloads[containerName]; set && m != nil {
+		memBytes = *m
+	} else if app.Spec.Memory.All != nil {
+		memBytes = *app.Spec.Memory.All
+	} else if container.Memory != nil {
+		memBytes = *container.Memory
+	}
+
+	var maximum int64
+	if cfg.WorkloadMemoryMaximum != nil {
+		maximum = *cfg.WorkloadMemoryMaximum
+	}
+
+	// For bytes, 0 is equivalent to "unlimited"
+	if maximum != 0 {
+		if memBytes > maximum {
+			return nil, fmt.Errorf(
+				"workload %v exceeds the workload-memory-maximum by %v",
+				containerName,
+				resource.NewQuantity(memBytes-maximum, resource.BinarySI).String())
+		} else if memBytes == 0 {
+			return nil, fmt.Errorf(
```

Comment:
What are your thoughts on just setting `memBytes = maximum` in this case?

Reply:
I see the merit in this since it helps the user get to their (probably) desired state without having to change anything. 

My main concern - what happens if the user is expecting to never run into memory issues by setting their limit to 0 and suddenly their app starts getting `OOM`'d due to the maximum be set as their value? This current approach prevents them from ever getting to that state since their desired state cannot be guaranteed.

Reply:
Resolved in offline conversation. We're going with your original approach and will update the docs to reflect the phrasing of "Setting memory to 0 will give that Acorn the maximum memory it is allowed to take".